Client/Frontend/Browser/BrowserViewController.swift

```diff
@@ -1665,3 +1669,63 @@ extension BrowserViewController: IntroViewControllerDelegate {
         })
     }
 }
+
+extension BrowserViewController: ContextMenuHelperDelegate {
+    func contextMenuHelper(contextMenuHelper: ContextMenuHelper, didLongPressElements elements: ContextMenuHelper.Elements) {
+        let actionSheetController = UIAlertController(title: nil, message: nil, preferredStyle: UIAlertControllerStyle.ActionSheet)
+        var dialogTitle: String?
+
+        if let url = elements.link {
+            dialogTitle = url.absoluteString
+            let newTabTitle = NSLocalizedString("Open In New Tab", comment: "Context menu item for opening a link in a new tab")
+            let openNewTabAction =  UIAlertAction(title: newTabTitle, style: UIAlertActionStyle.Default) { (action: UIAlertAction!) in
+                let request =  NSURLRequest(URL: url)
+                let tab = self.tabManager.addTab(request: request)
+            }
+            actionSheetController.addAction(openNewTabAction)
+
+            let copyTitle = NSLocalizedString("Copy Link", comment: "Context menu item for copying a link URL to the clipboard")
+            let copyAction = UIAlertAction(title: copyTitle, style: UIAlertActionStyle.Default) { (action: UIAlertAction!) -> Void in
+                var pasteBoard = UIPasteboard.generalPasteboard()
+                pasteBoard.string = url.absoluteString
+            }
+            actionSheetController.addAction(copyAction)
+        }
+
+        if let url = elements.image {
+            if dialogTitle == nil {
+                dialogTitle = url.absoluteString
+            }
+
+            let saveImageTitle = NSLocalizedString("Save Image", comment: "Context menu item for saving an image")
+            let saveImageAction = UIAlertAction(title: saveImageTitle, style: UIAlertActionStyle.Default) { (action: UIAlertAction!) -> Void in
+                self.getImage(url) { UIImageWriteToSavedPhotosAlbum($0, nil, nil, nil) }
```

Comment:
Yeah, but it can't go cross-frame. We can inject a script into frame A and frame B, but frame A cannot directly interact with B (assuming they're on separate domains) due to SOP.